python/python/bystro/ancestry/train.py

```diff
@@ -308,6 +313,64 @@ def _calc_fst(variant_counts: pd.Series, samples: pd.DataFrame) -> float:
     return (p * (1 - p) - total) / (p * (1 - p))
 
 
+def load_1kgp_vcf(TKGP_VCF):
+    """This is a temporary placeholder method to load in 1kgp vcf 
+    filtered down to the same variants as the gnomad loadings for testing
+    using plink2"""
+    headervcf=pd.read_csv(TKGP_VCF,delimiter='\t',comment='#')
+    dosagevcf = headervcf.replace('0|0', 0)
+    dosagevcf = dosagevcf.replace('0|1', 1)
+    dosagevcf = dosagevcf.replace('1|0', 1)
+    dosagevcf = dosagevcf.replace('1|1', 2)
+    dosagevcf=dosagevcf.rename(columns={"#CHROM": "Chromosome", "POS": "Position"}, errors="raise")
+    return dosagevcf
+    
+
+def load_pca_loadings(loadings, dosage_vcf):
+    """This loads in the gnomad PCs, reformats, and filters them to match
+    the format and variant list of the 1kgp vcf"""
+    loadings[['Chromosome', 'Position']] = loadings['locus'].str.split(':', expand=True)
+    #Match variant format with vcf
+    loadings['variant']=loadings.apply(lambda x: ':'.join([x['locus'][3:],x['alleles'][2],x['alleles'][6]]),axis=1)
+    #Remove variants that are missing from IGSR version of 1kgp
+    missing_variants = set(loadings.variant) - set(dosagevcf.ID)
+    loadings=loadings[~loadings['variant'].isin(missing_variants)]
+    #Remove brackets
+    loadings['loadings'] = loadings['loadings'].str[1:]
+    loadings['loadings'] = loadings['loadings'].str[:-1]
+    #Split PCs and join back
+    gnomadPCs = loadings['loadings'].str.split(',', expand=True)
+    loadings=loadings.join(gnomadPCs)
+    loadings=loadings.reset_index()
+    return loadings
+
+
+def apply_pca_transform(loadings, dosagevcf):
+    """Transforms vcf with genotypes in dosage format with PCs loadings from gnomad PCA"""
+    #Both files need to be np arrays for pca transformation
+    pc_loadings = loadings.iloc[:, 8:28]
+    genos = dosagevcf.iloc[:, 9:]
+    scaler = StandardScaler()
+    TGP_scaled = scaler.fit_transform(genos).T
+    TGP_scaled = np.array(TGP_scaled)
+    pc_loadings_array = pc_loadings.values.astype(float)
+    #Ensure the number of features matches the original dataset
+    if TGP_scaled.shape[1] != pc_loadings.shape[0]:
+        raise ValueError("Number of features in the new data doesn't match the number of features in the PCA loadings.")
+    #Scale 1kgp vcf
+    scaler = StandardScaler()
+    TGP_scaled = scaler.fit_transform(Genos).T
+    #Apply the loadings to the scaled 1kGP data
+    transformed_data = np.dot(TGP_scaled, pc_loadings_array)
+    #Add the IDs back on to PCs
+    IDs = genos.iloc[1]
```

Comment:
Much as with pie crust, the secret to manipulating dataframes (in my experience) is to do as little work as possible :)  If we have a genotype dataframe `genotypes` with dimensions (samples, variants) and a PCA loadings dataframe `loadings` with dimensions (variants, pc_dimensions), then we can just multiply them with the dot method of pd.DataFrame and get a type-checked matrix multiplication.  For example:

```
genotypes = pd.DataFrame(np.random.random(size=(5, 3)), columns=['variant1', 'variant2', 'variant3'], index=['alice', 'bob', 'carol', 'dave', 'eve'])
loadings = pd.DataFrame([[1, 2], [3, 4], [5, 6]], index=['variant1', 'variant2', 'variant3'], columns=['pc1', 'pc2'])
transformed_genotypes = genotypes.dot(loadings)

assert (transformed_genotypes.index == genotypes.index).all()
assert (transformed_genotypes.columns == loadings.columns).all()
```

The virtue of this approach is that if we try to multiply genotypes and loadings with mismatching sets of variants, we'll get a ValueError, and if we pass the right variants in the wrong order, `dot` will join them appropriately under the hood so that we get the right answer either way.